### ISSUE TITLE
Ability to rename request headers

### ISSUE THREAD

**Issue by ascheglov** [envoyproxy/envoy#8947]
Now that we can use request headers in the `request_headers_to_add` option, one might want to rename a header:

```
    request_headers_to_remove:
    - x-header
    request_headers_to_add:
    - header:
        key: x-new-name
        value: "%REQ(x-header)%"
```

However this will not work, because `HeaderParser::evaluateHeaders` removes headers first.

```
void HeaderParser::evaluateHeaders(Http::HeaderMap& headers,
                                   const StreamInfo::StreamInfo& stream_info) const {
  // Removing headers in the headers_to_remove_ list first makes
  // remove-before-add the default behavior as expected by users.
```

This was introduced in #5052 : "reverses the order of header evaluation as a solution to deduplicating request/response headers".

There are multiple ways to approach this header-renaming issue.

One is to add `request_headers_to_post_remove` which will be used after `request_headers_to_add`.

Another would be `request_headers_to_pre_add` but "post-remove" is better because headers-to-remove is just a list of names, when headers-to-add is much bulkier.

The most trivial option is to introduce a `map<string, string> rename_headers`.

There is also an option to add a `preserve_request_headers` list, to store original request headers somewhere in `StreamInfo`, and add a new formatter, e.g. `%PREQ(header-name)%` to access them. It might be useful because such headers won't be removed or modified by filters. However for now I just want to rename a couple of headers and this would be an overkill.

So, how about a new post-remove header list?

**Comment by zuercher** [comment]
I think we're somewhat constrained to how we can change this for API v2, but for API v3 I think we should consider a more generic way of defining header modifications. Instead of having 2 or 3 or more fields that manipulate headers I think we should have (roughly):

```
message HeaderModification {
  type.matcher.StringMatcher name = 1;

  oneof operation {
    string append = 2;
    string replace = 3;
    bool remove = 4;
    string rename = 5;
  }
}
repeated HeaderOperation header_modifications = 100;
```
(Obviously the message would not be inline and there are a number of validations that I've omitted for brevity.)

I think this lets us capture all the existing functionality plus renames, with the bonus of making it easier to add new features around headers in v3. I can imagine someone wanting to perform name transformations or gate operations on a string matcher for the header value in addition to the name.

@htuch what do you think?

**Comment by htuch** [comment]
@zuercher I agree; I'm going to tag this as v3 related. For v2, I feel a rename addition would be fine.
